api/graphite.go

```diff
@@ -698,6 +744,31 @@ func (s *Server) executePlan(ctx context.Context, orgId uint32, plan expr.Plan)
 	return out, err
 }
 
+// find the best consolidation method based on what was requested and what aggregations are available.
+func closestAggMethod(requested consolidation.Consolidator, available []conf.Method) consolidation.Consolidator {
+	// if there is only 1 consolidation method available, then that is all we can return.
+	if len(available) == 1 {
+		return consolidation.Consolidator(available[0])
+	}
+
+	avail := map[consolidation.Consolidator]struct{}{}
+	for _, a := range available {
+		avail[consolidation.Consolidator(a)] = struct{}{}
+	}
+	var orderOfPreference []consolidation.Consolidator
+	orderOfPreference, ok := rollupPreference[requested]
+	if !ok {
+		return consolidation.Consolidator(available[0])
```

Comment:
hmm this is interesting. 
conf/aggregations.go says that you can enable the following rollup methods: min, max, sum, avg, last
mdata/aggregator.go says generated rollups can be min, max, sum, cnt, lst (e.g. this adds count rollup which you cannot configure as a rollup, but you get it if you enable avg)
we have defined preferences for min max sum avg last, but not count
this means user can ask for consolidateBy('cnt'), and if the series has cnt listed in aggregations, but not as the first one, they will get a subpar result.
so we should probably add cnt to orderOfPreference?

Furthermore, since the requested consolidator is set via https://github.com/grafana/metrictank/blob/a7c8203d967fcf25ba463cede3e5310ba07a3181/consolidation/consolidation.go#L104-L130 , user can request any possible Consolidator value (incl range, stdev, etc)
in such case we should always return avg if we have it, whereas the aggregations list may not have avg as the first one (e.g. we may return cnt here if that was the first one defined, which would be incorrect)

Reply:
my 2nd point is invalid, because we make it clear in storage-aggregation.conf that we prefer to read from the first one listed, and so they should generally list avg first

Reply:
my 2nd point might be invalid,we make it clear in storage-aggregation.conf that we prefer to read from the first one listed, and so they should generally list avg first. i need to think a bit more about this

Reply:
All of your arguments are invalid.

https://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.consolidateBy

> Valid function names are ‘sum’, ‘average’, ‘min’, ‘max’, ‘first’ & ‘last’.

if `cnt` was a valid function for consolidateBy(), it wouldnt matter what rollup was used.  During runtime consolidation the "value" would just be the count of datapoints in the consolidation window.   Whether you use avg, min, sum, or cnt rollups, the number of points that get consolidate together doesnt change.   Using `consolidateBy(cnt)` would also be hugely confusing, as just because consolidateBy is passed, doesnt mean that any runtime consolidation will happen.  If a user starts with a narrow time window then zooms out to a range that requires runtime consolidation the values would change dramatically.  And this is probably why `cnt` is not a valid consolidation method for consolidateBy

Returning the "default" rollup when a consolidation is requested for which there is no rollup is the correct behaviour.